File: mongodb-faker-generator/python/generate_fake_data.py
"""Utility to build and seed a sample MongoDB dataset for demos/tests."""

from faker import Faker
from pymongo import MongoClient, ASCENDING, DESCENDING
from datetime import datetime
import random
from typing import List, Dict, Any, Callable

# Initialize Faker and MongoDB
fake = Faker("pt_PT")
client = MongoClient("mongodb://localhost:27017/")  # Adjust connection string as needed
db = client["database_demo"]  # Database name


def _faker_attr(name: str, fallback: Callable[[], str]) -> str:
    """
    Call a Faker provider if it exists, otherwise fall back.
    """
    provider = getattr(fake, name, None)
    if callable(provider):
        return provider()
    return fallback()


def generate_users(n: int = 100) -> List[Dict[str, Any]]:
    """
    Generate complex user documents containing profile, account, and metadata info.

    Parameters
    ----------
    n : int, optional
        Number of user records to create.

    Returns
    -------
    list[dict]
        List of Mongo-style user documents ready for insertion.
    """
    users = []

    for _ in range(n):
        birth_date = fake.date_of_birth(minimum_age=18, maximum_age=79)
        birth_datetime = datetime.combine(birth_date, datetime.min.time())
        user = {
            "_id": fake.uuid4(),
            "username": fake.user_name(),
            "email": fake.email(),
            "profile": {
                "firstName": fake.first_name(),
                "lastName": fake.last_name(),
                "fullName": fake.name(),
                "avatar": fake.image_url(),
                "bio": fake.text(max_nb_chars=200),
                "birthDate": birth_datetime,
                "gender": fake.random_element(["M", "F", "Other", None]),
                "phone": fake.phone_number(),
                "address": {
                    "street": fake.street_address(),
                    "city": fake.city(),
                    "state": _faker_attr("state", fake.city),
                    "country": "Portugal",
                    "postalCode": fake.postcode(),
                    "coordinates": {
                        "type": "Point",
                        "coordinates": [
                            float(fake.longitude()),
                            float(fake.latitude()),
                        ],
                    },
                },
            },
            "account": {
                "type": fake.random_element(["free", "premium", "enterprise"]),
                "status": fake.random_element(["active", "inactive", "suspended"]),
                "createdAt": fake.date_time_between(start_date="-3y", end_date="now"),
                "lastLogin": fake.date_time_between(start_date="-30d", end_date="now"),
                "loginCount": fake.random_int(min=0, max=1000),
                "preferences": {
                    "newsletter": fake.boolean(chance_of_getting_true=70),
                    "notifications": fake.boolean(chance_of_getting_true=80),
                    "language": fake.random_element(["pt", "en", "es"]),
                    "theme": fake.random_element(["light", "dark", "auto"]),
                },
            },
            "tags": fake.words(nb=random.randint(2, 6)),
            "metadata": {
                "source": fake.random_element(["web", "mobile", "api"]),
                "ipAddress": fake.ipv4(),
                "userAgent": fake.user_agent(),
            },
        }
        users.append(user)

    return users


def generate_products(n: int = 500) -> List[Dict[str, Any]]:
    """
    Generate product documents with nested pricing, inventory, and attribute fields.

    Parameters
    ----------
    n : int, optional
        Number of products to emit.
    """
    products = []
    categories = ["Electronics", "Books", "Clothing", "Home & Garden", "Sports", "Toys"]

    for _ in range(n):
        category = fake.random_element(categories)
        product = {
            "sku": fake.ean13(),
            "name": fake.catch_phrase(),
            "description": fake.paragraph(nb_sentences=5),
            "category": category,
            "subcategory": fake.word(),
            "brand": fake.company(),
            "price": {
                "amount": round(fake.random.uniform(10, 1000), 2),
                "currency": "EUR",
                "discount": round(fake.random.uniform(0, 30), 2)
                if fake.boolean(chance_of_getting_true=30)
                else 0,
            },
            "inventory": {
                "inStock": fake.boolean(chance_of_getting_true=85),
                "quantity": fake.random_int(min=0, max=1000),
                "warehouse": fake.random_element(["Porto", "Lisboa", "Faro"]),
            },
            "attributes": {
                "color": fake.color_name()
                if category in ["Clothing", "Home & Garden"]
                else None,
                "size": fake.random_element(["S", "M", "L", "XL"])
                if category == "Clothing"
                else None,
                "weight": f"{fake.random.uniform(0.1, 50):.2f} kg",
                "dimensions": {
                    "length": fake.random_int(min=10, max=200),
                    "width": fake.random_int(min=10, max=150),
                    "height": fake.random_int(min=5, max=100),
                    "unit": "cm",
                },
            },
            "ratings": {
                "average": round(fake.random.uniform(1, 5), 1),
                "count": fake.random_int(min=0, max=5000),
            },
            "images": [fake.image_url() for _ in range(fake.random_int(1, 5))],
            "tags": fake.words(nb=random.randint(3, 8)),
            "createdAt": fake.date_time_between(start_date="-2y", end_date="now"),
            "updatedAt": fake.date_time_between(start_date="-30d", end_date="now"),
        }
        if not product["inventory"]["inStock"]:
            product["inventory"]["quantity"] = 0
        products.append(product)

    return products


def generate_transactions(
    users: List[Dict], products: List[Dict], n: int = 1000
) -> List[Dict[str, Any]]:
    """
    Generate transactional orders that reference previously created users/products.

    Parameters
    ----------
    users : list[dict]
        User documents used to resolve userId references.
    products : list[dict]
        Product documents used to pull pricing/SKU details.
    n : int, optional
        Number of transactions to produce.
    """
    transactions = []

    for _ in range(n):
        num_items = fake.random_int(min=1, max=5)
        items = []
        total = 0

        for _ in range(num_items):
            product = fake.random_element(products)
            quantity = fake.random_int(min=1, max=3)
            item_price = product["price"]["amount"]
            item_total = round(item_price * quantity, 2)
            items.append(
                {
                    "productSku": product["sku"],
                    "productName": product["name"],
                    "quantity": quantity,
                    "unitPrice": item_price,
                    "total": item_total,
                }
            )
            total += item_total

        transaction = {
            "orderId": fake.uuid4(),
            "userId": fake.random_element(users)["_id"],
            "status": fake.random_element(
                ["pending", "processing", "shipped", "delivered", "cancelled"]
            ),
            "items": items,
            "payment": {
                "method": fake.random_element(
                    ["credit_card", "debit_card", "paypal", "mbway", "bank_transfer"]
                ),
                "status": fake.random_element(["pending", "completed", "failed"]),
                "transactionId": fake.sha256()[:16],
            },
            "shipping": {
                "address": {
                    "street": fake.street_address(),
                    "city": fake.city(),
                    "postalCode": fake.postcode(),
                    "country": "Portugal",
                },
                "method": fake.random_element(["standard", "express", "overnight"]),
                "trackingNumber": fake.ean13()
                if fake.boolean(chance_of_getting_true=70)
                else None,
                "estimatedDelivery": datetime.combine(
                    fake.future_date(end_date="+30d"), datetime.min.time()
                ),
            },
            "totals": {
                "subtotal": round(total, 2),
                "tax": round(total * 0.23, 2),  # Portuguese VAT for realistic totals
                "shipping": 0.0,
                "total": 0,  # Placeholder, set below
            },
            "timestamps": {
                "created": fake.date_time_between(start_date="-1y", end_date="now"),
                "updated": fake.date_time_between(start_date="-30d", end_date="now"),
            },
        }
        transaction["shipping"]["cost"] = round(fake.random.uniform(0, 20), 2)
        transaction["totals"]["shipping"] = transaction["shipping"]["cost"]
        subtotal = transaction["totals"]["subtotal"]
        tax = transaction["totals"]["tax"]
        shipping = transaction["totals"]["shipping"]
        transaction["totals"]["total"] = round(subtotal + tax + shipping, 2)

        if transaction["status"] == "delivered":
            transaction["payment"]["status"] = "completed"
        elif (
            transaction["payment"]["status"] == "failed"
            and transaction["status"] == "shipped"
        ):
            transaction["status"] = fake.random_element(["processing", "cancelled"])
        transactions.append(transaction)

    return transactions


def generate_logs(users: List[Dict], n: int = 5000) -> List[Dict[str, Any]]:
    """
    Generate application log entries for auth, API, and error events.

    Parameters
    ----------
    users : list[dict]
        Source user documents to associate optional userId values.
    n : int, optional
        Number of log entries to create.
    """
    level_cycle = ["INFO", "DEBUG", "WARNING", "ERROR", "CRITICAL", "INFO"]
    weighted_levels = ["INFO", "INFO", "DEBUG", "WARNING", "ERROR", "CRITICAL"]
    log_types = ["login", "logout", "page_view", "api_call", "error", "performance"]

    logs = []
    for idx in range(n):
        if idx < len(level_cycle):
            level = level_cycle[idx]
        else:
            level = fake.random_element(weighted_levels)
        log = {
            "timestamp": fake.date_time_between(start_date="-7d", end_date="now"),
            "level": level,
            "type": fake.random_element(log_types),
            "userId": fake.random_element(users)["_id"]
            if fake.boolean(chance_of_getting_true=80)
            else None,
            "sessionId": fake.uuid4(),
            "message": fake.sentence(),
            "metadata": {
                "ip": fake.ipv4(),
                "userAgent": fake.user_agent(),
                "endpoint": f"/api/{fake.word()}/{fake.word()}",
                "method": fake.random_element(["GET", "POST", "PUT", "DELETE"]),
                "statusCode": fake.random_element([200, 201, 400, 401, 403, 404, 500]),
                "responseTime": fake.random_int(min=10, max=3000),
            },
            "error": {
                "type": fake.word()
                if fake.boolean(chance_of_getting_true=20)
                else None,
                "stack": fake.paragraph()
                if fake.boolean(chance_of_getting_true=20)
                else None,
            },
        }
        logs.append(log)

    return logs


def insert_data_to_mongodb():
    """
    Regenerate every dataset, drop existing collections, insert data, and recreate indexes.
    """
    print("🔄 Generating fake data...")

    # Generate data
    users = generate_users(100)
    products = generate_products(500)
    transactions = generate_transactions(users, products, 1000)
    logs = generate_logs(users, 5000)

    print("📊 Data generated:")
    print(f"  - Users: {len(users)}")
    print(f"  - Products: {len(products)}")
    print(f"  - Transactions: {len(transactions)}")
    print(f"  - Logs: {len(logs)}")

    # Insert into MongoDB
    print("\n💾 Inserting into MongoDB...")

    # Users collection
    users_collection = db["users"]
    users_collection.drop()  # Clean existing data
    users_collection.insert_many(users)
    users_collection.create_index([("email", ASCENDING)], unique=True)
    users_collection.create_index([("username", ASCENDING)], unique=True)
    users_collection.create_index(
        [("profile.address.coordinates", "2dsphere")]
    )  # Geospatial index
    print("✅ Users inserted")

    # Products collection
    products_collection = db["products"]
    products_collection.drop()
    products_collection.insert_many(products)
    products_collection.create_index([("sku", ASCENDING)], unique=True)
    products_collection.create_index([("category", ASCENDING)])
    products_collection.create_index([("price.amount", ASCENDING)])
    products_collection.create_index(
        [("name", "text"), ("description", "text")]
    )  # Text search
    print("✅ Products inserted")

    # Transactions collection
    transactions_collection = db["transactions"]
    transactions_collection.drop()
    transactions_collection.insert_many(transactions)
    transactions_collection.create_index([("orderId", ASCENDING)], unique=True)
    transactions_collection.create_index([("userId", ASCENDING)])
    transactions_collection.create_index([("timestamps.created", DESCENDING)])
    print("✅ Transactions inserted")

    # Logs collection with TTL
    logs_collection = db["logs"]
    logs_collection.drop()
    logs_collection.insert_many(logs)
    logs_collection.create_index([("timestamp", DESCENDING)])
    logs_collection.create_index([("userId", ASCENDING)])
    logs_collection.create_index([("level", ASCENDING)])
    # TTL index - automatically delete logs older than 30 days
    logs_collection.create_index(
        [("timestamp", ASCENDING)], expireAfterSeconds=30 * 24 * 60 * 60
    )
    print("✅ Logs inserted")

    print("\n📈 Database statistics:")
    stats = db.command("dbstats")
    print(f"  - Database: {stats['db']}")
    print(f"  - Collections: {stats['collections']}")
    print(f"  - Data Size: {stats['dataSize'] / 1024 / 1024:.2f} MB")


def query_examples():
    """
    Run several read/report examples so students can validate the dataset interactively.
    """
    print("\n🔍 Example Queries:")

    # Example 1: Find premium users from Porto
    result = db.users.find(
        {"account.type": "premium", "profile.address.city": "Porto"}
    ).limit(5)
    print(f"\n1. Premium users from Porto: {len(list(result))}")

    # Example 2: Products with high ratings
    result = db.products.find({"ratings.average": {"$gte": 4.5}}).limit(5)
    print(f"2. Products with rating >= 4.5: {len(list(result))}")

    # Example 3: Recent large transactions
    result = db.transactions.find(
        {
            "totals.total": {"$gte": 500},
            "timestamps.created": {
                "$gte": datetime.now().replace(
                    hour=0, minute=0, second=0, microsecond=0
                )
            },
        }
    )
    print(f"3. Today's transactions over €500: {len(list(result))}")

    # Example 4: Error logs count
    result = db.logs.count_documents({"level": "ERROR"})
    print(f"4. Error logs: {result}")

    # Example 5: Aggregation - Sales by category
    pipeline = [
        {
            "$lookup": {
                "from": "products",
                "localField": "items.productSku",
                "foreignField": "sku",
                "as": "productDetails",
            }
        },
        {"$unwind": "$productDetails"},
        {
            "$group": {
                "_id": "$productDetails.category",
                "totalSales": {"$sum": "$totals.total"},
                "count": {"$sum": 1},
            }
        },
        {"$sort": {"totalSales": -1}},
    ]
    result = list(db.transactions.aggregate(pipeline))
    print("\n5. Sales by category:")
    for cat in result[:3]:
        print(
            f"   - {cat['_id']}: €{cat['totalSales']:.2f} ({cat['count']} transactions)"
        )


if __name__ == "__main__":
    try:
        insert_data_to_mongodb()
        query_examples()
        print("\n✅ All done! Check your MongoDB database.")
    except Exception as e:
        print(f"\n❌ Error: {e}")
    finally:
        client.close()

File: mongodb-faker-generator/javascript/generate_fake_data.js
// Utility script to seed MongoDB with a realistic e-commerce workload using Faker.
import { faker } from '@faker-js/faker';
import { MongoClient } from 'mongodb';

// Configure Faker
faker.locale = 'pt_PT';

// MongoDB connection
const uri = 'mongodb://localhost:27017';
const client = new MongoClient(uri);
const dbName = 'database_demo';

/**
 * Build an array of mock user profiles with nested account/profile data.
 *
 * @param {number} [n=100] - Number of user documents to create.
 * @returns {Array<object>} Generated user documents ready for MongoDB.
 */
function generateUsers(n = 100) {
    const users = [];
    
    for (let i = 0; i < n; i++) {
        const user = {
            _id: faker.datatype.uuid(),
            username: faker.internet.userName(),
            email: faker.internet.email(),
            profile: {
                firstName: faker.person.firstName(),
                lastName: faker.person.lastName(),
                fullName: faker.person.fullName(),
                avatar: faker.image.avatar(),
                bio: faker.lorem.paragraph({ min: 2, max: 4 }),
                birthDate: faker.date.birthdate({ min: 18, max: 80, mode: 'age' }),
                gender: faker.helpers.arrayElement(['M', 'F', 'Other', null]),
                phone: faker.phone.number(),
                address: {
                    street: faker.location.streetAddress(),
                    city: faker.location.city(),
                    state: faker.location.state(),
                    country: 'Portugal',
                    postalCode: faker.location.zipCode(),
                    coordinates: {
                        type: 'Point',
                        coordinates: [
                            parseFloat(faker.location.longitude()),
                            parseFloat(faker.location.latitude())
                        ]
                    }
                }
            },
            account: {
                type: faker.helpers.arrayElement(['free', 'premium', 'enterprise']),
                status: faker.helpers.arrayElement(['active', 'inactive', 'suspended']),
                createdAt: faker.date.between({ from: new Date('2021-01-01'), to: new Date() }),
                lastLogin: faker.date.recent({ days: 30 }),
                loginCount: faker.number.int({ min: 0, max: 1000 }),
                preferences: {
                    newsletter: faker.datatype.boolean({ probability: 0.7 }),
                    notifications: faker.datatype.boolean({ probability: 0.8 }),
                    language: faker.helpers.arrayElement(['pt', 'en', 'es']),
                    theme: faker.helpers.arrayElement(['light', 'dark', 'auto'])
                }
            },
            tags: faker.helpers.multiple(() => faker.word.noun(), { count: { min: 2, max: 6 } }),
            metadata: {
                source: faker.helpers.arrayElement(['web', 'mobile', 'api']),
                ipAddress: faker.internet.ipv4(),
                userAgent: faker.internet.userAgent()
            }
        };
        users.push(user);
    }
    
    return users;
}

/**
 * Build an array of product documents spanning multiple categories.
 *
 * @param {number} [n=500] - Number of products to generate.
 * @returns {Array<object>} Product payloads with pricing, inventory, etc.
 */
function generateProducts(n = 500) {
    const products = [];
    const categories = ['Electronics', 'Books', 'Clothing', 'Home & Garden', 'Sports', 'Toys'];
    
    for (let i = 0; i < n; i++) {
        const category = faker.helpers.arrayElement(categories);
        const product = {
            sku: faker.commerce.isbn(),
            name: faker.commerce.productName(),
            description: faker.commerce.productDescription(),
            category: category,
            subcategory: faker.commerce.department(),
            brand: faker.company.name(),
            price: {
                amount: parseFloat(faker.commerce.price({ min: 10, max: 1000, dec: 2 })),
                currency: 'EUR',
                discount: faker.datatype.boolean({ probability: 0.3 }) 
                    ? parseFloat(faker.commerce.price({ min: 0, max: 30, dec: 2 })) 
                    : 0
            },
            inventory: {
                inStock: faker.datatype.boolean({ probability: 0.85 }),
                quantity: faker.number.int({ min: 0, max: 1000 }),
                warehouse: faker.helpers.arrayElement(['Porto', 'Lisboa', 'Faro'])
            },
            attributes: {
                color: ['Clothing', 'Home & Garden'].includes(category) 
                    ? faker.color.human() 
                    : null,
                size: category === 'Clothing' 
                    ? faker.helpers.arrayElement(['S', 'M', 'L', 'XL']) 
                    : null,
                weight: `${faker.number.float({ min: 0.1, max: 50, precision: 0.01 })} kg`,
                dimensions: {
                    length: faker.number.int({ min: 10, max: 200 }),
                    width: faker.number.int({ min: 10, max: 150 }),
                    height: faker.number.int({ min: 5, max: 100 }),
                    unit: 'cm'
                }
            },
            ratings: {
                average: parseFloat(faker.number.float({ min: 1, max: 5, precision: 0.1 })),
                count: faker.number.int({ min: 0, max: 5000 })
            },
            images: faker.helpers.multiple(() => faker.image.url(), { 
                count: faker.number.int({ min: 1, max: 5 }) 
            }),
            tags: faker.helpers.multiple(() => faker.word.noun(), { count: { min: 3, max: 8 } }),
            createdAt: faker.date.between({ from: new Date('2022-01-01'), to: new Date() }),
            updatedAt: faker.date.recent({ days: 30 })
        };
        products.push(product);
    }
    
    return products;
}

/**
 * Generate purchase transactions that reference the supplied users/products.
 *
 * @param {Array<object>} users - Previously generated user docs (for userId refs).
 * @param {Array<object>} products - Product docs for SKU references.
 * @param {number} [n=1000] - Number of transactions to create.
 * @returns {Array<object>} Transaction documents with embedded items.
 */
function generateTransactions(users, products, n = 1000) {
    const transactions = [];
    
    for (let i = 0; i < n; i++) {
        const numItems = faker.number.int({ min: 1, max: 5 });
        const items = [];
        let subtotal = 0;
        
        for (let j = 0; j < numItems; j++) {
            const product = faker.helpers.arrayElement(products);
            const quantity = faker.number.int({ min: 1, max: 3 });
            const itemTotal = product.price.amount * quantity;
            
            items.push({
                productSku: product.sku,
                productName: product.name,
                quantity: quantity,
                unitPrice: product.price.amount,
                total: parseFloat(itemTotal.toFixed(2))
            });
            subtotal += itemTotal;
        }
        
        // Apply Portuguese VAT and a random shipping fee to mirror real totals.
        const tax = subtotal * 0.23;
        const shipping = parseFloat(faker.commerce.price({ min: 0, max: 20, dec: 2 }));
        
        const transaction = {
            orderId: faker.datatype.uuid(),
            userId: faker.helpers.arrayElement(users)._id,
            status: faker.helpers.arrayElement(['pending', 'processing', 'shipped', 'delivered', 'cancelled']),
            items: items,
            payment: {
                method: faker.helpers.arrayElement(['credit_card', 'debit_card', 'paypal', 'mbway', 'bank_transfer']),
                status: faker.helpers.arrayElement(['pending', 'completed', 'failed']),
                transactionId: faker.git.commitSha().substring(0, 16)
            },
            shipping: {
                address: {
                    street: faker.location.streetAddress(),
                    city: faker.location.city(),
                    postalCode: faker.location.zipCode(),
                    country: 'Portugal'
                },
                method: faker.helpers.arrayElement(['standard', 'express', 'overnight']),
                trackingNumber: faker.datatype.boolean({ probability: 0.7 }) 
                    ? faker.string.alphanumeric(13).toUpperCase() 
                    : null,
                estimatedDelivery: faker.date.future({ years: 0.1 })
            },
            totals: {
                subtotal: parseFloat(subtotal.toFixed(2)),
                tax: parseFloat(tax.toFixed(2)),
                shipping: parseFloat(shipping),
                total: parseFloat((subtotal + tax + parseFloat(shipping)).toFixed(2))
            },
            timestamps: {
                created: faker.date.between({ from: new Date('2023-01-01'), to: new Date() }),
                updated: faker.date.recent({ days: 30 })
            }
        };
        transactions.push(transaction);
    }
    
    return transactions;
}

/**
 * Generate synthetic application logs optionally tied to user ids.
 *
 * @param {Array<object>} users - User docs to pick IDs from.
 * @param {number} [n=5000] - Number of log entries to emit.
 * @returns {Array<object>} Log documents spanning multiple levels/types.
 */
function generateLogs(users, n = 5000) {
    const logs = [];
    const logLevels = ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'];
    const logTypes = ['login', 'logout', 'page_view', 'api_call', 'error', 'performance'];
    
    for (let i = 0; i < n; i++) {
        const hasError = faker.datatype.boolean({ probability: 0.2 });
        
        const log = {
            timestamp: faker.date.recent({ days: 7 }),
            level: faker.helpers.arrayElement(logLevels),
            type: faker.helpers.arrayElement(logTypes),
            userId: faker.datatype.boolean({ probability: 0.8 }) 
                ? faker.helpers.arrayElement(users)._id 
                : null,
            sessionId: faker.datatype.uuid(),
            message: faker.hacker.phrase(),
            metadata: {
                ip: faker.internet.ipv4(),
                userAgent: faker.internet.userAgent(),
                endpoint: `/api/${faker.word.noun()}/${faker.word.verb()}`,
                method: faker.helpers.arrayElement(['GET', 'POST', 'PUT', 'DELETE']),
                statusCode: faker.helpers.arrayElement([200, 201, 400, 401, 403, 404, 500]),
                responseTime: faker.number.int({ min: 10, max: 3000 })
            },
            error: hasError ? {
                type: faker.hacker.abbreviation(),
                stack: faker.lorem.paragraphs({ min: 2, max: 5 })
            } : null
        };
        logs.push(log);
    }
    
    return logs;
}

// Main function to insert data
/**
 * Connect to MongoDB, regenerate datasets, insert them, and run sample queries.
 *
 * @returns {Promise<void>}
 */
async function insertDataToMongoDB() {
    try {
        // Connect to MongoDB
        await client.connect();
        console.log('✅ Connected to MongoDB');
        
        const db = client.db(dbName);
        
        console.log('🔄 Generating fake data...');
        
        // Generate data
        const users = generateUsers(100);
        const products = generateProducts(500);
        const transactions = generateTransactions(users, products, 1000);
        const logs = generateLogs(users, 5000);
        
        console.log('📊 Data generated:');
        console.log(`  - Users: ${users.length}`);
        console.log(`  - Products: ${products.length}`);
        console.log(`  - Transactions: ${transactions.length}`);
        console.log(`  - Logs: ${logs.length}`);
        
        console.log('\n💾 Inserting into MongoDB...');
        
        // Drop existing collections
        await db.collection('users').drop().catch(() => {});
        await db.collection('products').drop().catch(() => {});
        await db.collection('transactions').drop().catch(() => {});
        await db.collection('logs').drop().catch(() => {});
        
        // Insert users
        const usersCollection = db.collection('users');
        await usersCollection.insertMany(users);
        await usersCollection.createIndex({ email: 1 }, { unique: true });
        await usersCollection.createIndex({ username: 1 }, { unique: true });
        await usersCollection.createIndex({ 'profile.address.coordinates': '2dsphere' });
        console.log('✅ Users inserted');
        
        // Insert products
        const productsCollection = db.collection('products');
        await productsCollection.insertMany(products);
        await productsCollection.createIndex({ sku: 1 }, { unique: true });
        await productsCollection.createIndex({ category: 1 });
        await productsCollection.createIndex({ 'price.amount': 1 });
        await productsCollection.createIndex({ name: 'text', description: 'text' });
        console.log('✅ Products inserted');
        
        // Insert transactions
        const transactionsCollection = db.collection('transactions');
        await transactionsCollection.insertMany(transactions);
        await transactionsCollection.createIndex({ orderId: 1 }, { unique: true });
        await transactionsCollection.createIndex({ userId: 1 });
        await transactionsCollection.createIndex({ 'timestamps.created': -1 });
        console.log('✅ Transactions inserted');
        
        // Insert logs with TTL
        const logsCollection = db.collection('logs');
        await logsCollection.insertMany(logs);
        await logsCollection.createIndex({ timestamp: -1 });
        await logsCollection.createIndex({ userId: 1 });
        await logsCollection.createIndex({ level: 1 });
        await logsCollection.createIndex(
            { timestamp: 1 }, 
            { expireAfterSeconds: 30 * 24 * 60 * 60 } // 30 days TTL
        );
        console.log('✅ Logs inserted');
        
        // Show database stats
        const stats = await db.stats();
        console.log('\n📈 Database statistics:');
        console.log(`  - Database: ${stats.db}`);
        console.log(`  - Collections: ${stats.collections}`);
        console.log(`  - Data Size: ${(stats.dataSize / 1024 / 1024).toFixed(2)} MB`);
        
        // Example queries
        await runExampleQueries(db);
        
    } catch (error) {
        console.error('❌ Error:', error);
    } finally {
        await client.close();
        console.log('\n👋 MongoDB connection closed');
    }
}

/**
 * Print a handful of business-driven queries/aggregations to validate the dataset.
 *
 * @param {import("mongodb").Db} db - Active database connection.
 * @returns {Promise<void>}
 */
async function runExampleQueries(db) {
    console.log('\n🔍 Example Queries:');
    
    // 1. Premium users from Porto
    const premiumUsers = await db.collection('users')
        .find({
            'account.type': 'premium',
            'profile.address.city': 'Porto'
        })
        .limit(5)
        .toArray();
    console.log(`\n1. Premium users from Porto: ${premiumUsers.length}`);
    
    // 2. High-rated products
    const highRatedProducts = await db.collection('products')
        .find({
            'ratings.average': { $gte: 4.5 }
        })
        .limit(5)
        .toArray();
    console.log(`2. Products with rating >= 4.5: ${highRatedProducts.length}`);
    
    // 3. Recent large transactions
        const today = new Date();
        today.setHours(0, 0, 0, 0);
    
    const largeTransactions = await db.collection('transactions')
        .find({
            'totals.total': { $gte: 500 },
            'timestamps.created': { $gte: today }
        })
        .toArray();
    console.log(`3. Today's transactions over €500: ${largeTransactions.length}`);
    
    // 4. Error logs count
    const errorCount = await db.collection('logs')
        .countDocuments({ level: 'ERROR' });
    console.log(`4. Error logs: ${errorCount}`);
    
    // 5. Sales by category (aggregation)
    const salesByCategory = await db.collection('transactions')
        .aggregate([
            {
                $unwind: '$items'
            },
            {
                $lookup: {
                    from: 'products',
                    localField: 'items.productSku',
                    foreignField: 'sku',
                    as: 'product'
                }
            },
            {
                $unwind: '$product'
            },
            {
                $group: {
                    _id: '$product.category',
                    totalSales: { $sum: '$items.total' },
                    count: { $sum: 1 }
                }
            },
            {
                $sort: { totalSales: -1 }
            },
            {
                $limit: 3
            }
        ])
        .toArray();
    
    console.log('\n5. Sales by category:');
    salesByCategory.forEach(cat => {
        console.log(`   - ${cat._id}: €${cat.totalSales.toFixed(2)} (${cat.count} items)`);
    });
}

// Run the script
insertDataToMongoDB();
